### PR TITLE
hyperspeedcube: init at 1.0.6

### PR DESCRIPTION
**Pull request by OmnipotentEntity**
## Description of changes

New Package, hyperspeedcube.  Packaged because getting it to work properly was kind of involved, and I wanted to save others the work.

Clears the wayland session to force initialization using X11.  The application should support Wayland, but in practice it wound up crashing pretty quickly with my setup.  It also clears the XDG_SESSION_TYPE due to some logic in the keybindings that crash if it's running in X11 under a session that it detects "should be" Wayland.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/games/hyperspeedcube/default.nix [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  postInstall = ''
+    patchelf \
+      --add-needed ${vulkan-loader}/lib/libvulkan.so.1 \
+      --add-needed ${libGL}/lib/libEGL.so.1 \
+      $out/bin/hyperspeedcube
+    wrapProgram $out/bin/hyperspeedcube --set WAYLAND_DISPLAY "" --set XDG_SESSION_TYPE ""
+    touch $out/bin/nonportable
+  '';
+
+  meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/games/hyperspeedcube/default.nix:116]:
```suggestion
  meta = {
```

File: pkgs/games/hyperspeedcube/default.nix [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    description = ''
```
Comment (AndersonTorres) [inline pkgs/games/hyperspeedcube/default.nix:117]:
```suggestion
    longDescription = ''
```

Also, write a short, one-line `description`  according to the guidelines.

File: pkgs/games/hyperspeedcube/default.nix [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      Hyperspeedcube is a modern, beginner-friendly 3D and 4D Rubik's cube
+      simulator with customizable mouse and keyboard controls and advanced
+      features for speedsolving. It's been used to break numerous speedsolving
+      records and runs on all major operating systems plus the web.'';
```
Comment (AndersonTorres) [inline pkgs/games/hyperspeedcube/default.nix:121]:
```suggestion
      records and runs on all major operating systems plus the web.
    '';
```

File: pkgs/games/hyperspeedcube/default.nix [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    homepage    = "https://ajfarkas.dev/hyperspeedcube/";
+    license     = licenses.cc-by-nc-sa-40;
+    maintainers = [ maintainers.omnipotententity ];
+    platforms   = [ "x86_64-linux" ];
```
Comment (AndersonTorres) [inline pkgs/games/hyperspeedcube/default.nix:125]:
```suggestion
    homepage = "https://ajfarkas.dev/hyperspeedcube/";
    license = lib.licenses.cc-by-nc-sa-40;
    maintainers = [ lib.maintainers.omnipotententity ];
    platforms = [ "x86_64-linux" ];
```

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -37020,6 +37020,8 @@ with pkgs;
 
   hyperrogue = callPackage ../games/hyperrogue { };
 
+  hyperspeedcube = callPackage ../games/hyperspeedcube { };
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:37023]: Use the by-name hierarchy.

File: pkgs/games/hyperspeedcube/default.nix [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    libGL
+
+    # WINIT_UNIX_BACKEND=wayland
+    wayland
+
+    # WINIT_UNIX_BACKEND=x11
+    xorg.libXcursor
+    xorg.libXrandr
+    xorg.libXi
+    xorg.libX11
```
Comment (AndersonTorres) [inline pkgs/games/hyperspeedcube/default.nix:95]:
```suggestion
    libXcursor
    libXrandr
    libXi
    libX11
```

File: pkgs/games/hyperspeedcube/default.nix [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+     "key-names-1.5.1" = "sha256-ZO3mlCoUHCaMnxD2vQno6GNx0+JmwZ8OiqpyveiVfvc=";
+    };
+  };
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+    xorg.libxcb.dev
```
Comment (AndersonTorres) [inline pkgs/games/hyperspeedcube/default.nix:61]:
```suggestion
    (lib.getDev libxcb)
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by OmnipotentEntity** [comment]
I wasn't aware of the new `by-name` hierarchy.  Addressed this and the rest of the things mentioned.

**Comment by OmnipotentEntity** [comment]
Most recent push only removed some stray comments that were not applicable to this package.

File: pkgs/by-name/hy/hyperspeedcube/package.nix [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+, shaderc
+, stdenv
+, vulkan-extension-layer
+, vulkan-headers
+, vulkan-loader
+, vulkan-tools
+, vulkan-tools-lunarg
+, vulkan-validation-layers
+, wayland
+, wrapGAppsHook
```
Comment (isabelroses) [inline pkgs/by-name/hy/hyperspeedcube/package.nix:37]: `wrapGAppsHook` should be either `wrapGAppsHook3` for gtk3 apps or `wrapGAppsHook4` for gtk4 apps.

File: pkgs/by-name/hy/hyperspeedcube/package.nix [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    patchelf \
+      --add-needed ${vulkan-loader}/lib/libvulkan.so.1 \
+      --add-needed ${libGL}/lib/libEGL.so.1 \
+      $out/bin/hyperspeedcube
+    wrapProgram $out/bin/hyperspeedcube --set WAYLAND_DISPLAY "" --set XDG_SESSION_TYPE ""
+    touch $out/bin/nonportable
+  '';
+
+  meta = {
+    description = "Hyperspeedcube is a 3D and 4D Rubik's cube simulator.";
```
Comment (isabelroses) [inline pkgs/by-name/hy/hyperspeedcube/package.nix:119]: [nixos.org/manual/nixpkgs/unstable/#var-meta-description](https://nixos.org/manual/nixpkgs/unstable/#var-meta-description)

> Don’t include a period at the end.
```suggestion
    description = "Hyperspeedcube is a 3D and 4D Rubik's cube simulator";
```

**Review by isabelroses — APPROVED** [review]

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 307226` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>hyperspeedcube</li>
  </ul>
</details>
runs fine, LGTM

**Pull request merged**